Migrate `Quick Start - List Item` refresh logic to `ImprovedMySiteFragment`

ashiagr opened the pull request:
Issue #15127

This PR 
- builds `Quick Start - List Item` in `ImprovedMySiteFragment` using the QS repository's categories. This is done so that the list item can be refreshed every time the categories are refreshed from the QS repository.
- refreshes the `Quick Start - List Item` in `ImprovedMySiteFragment` when the `QuickStartFullScreenDialogFragment` is dismissed.
- handles `QuickStartFullScreenDialogFragment`'s confirm action which is triggered when a [quick start task is tapped from this fragment](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/quickstart/QuickStartFullScreenDialogFragment.java#L157) by calling the QS repository's `onQuickStartTaskCardClick` via the VM. QS focus points and snack bars are taken care of automatically (note that existing snack bars from `ImprovedMySiteFragment` are reused). 

To test:

1. Go to `My Site` -> `App Settings` -> `Test feature configuration`.
2. Disable `QuickStartDynamicCardsFeatureConfig`, enable `MySiteImprovementsFeatureConfig`
3. Restart the app
4. Select a site with quick start in progress 
5. Click quick start task type list item
6. Notice that `QuickStartFullScreenDialogFragment` gets opened
7. Select a task from the `QuickStartFullScreenDialogFragment`
8. Notice that when `QuickStartFullScreenDialogFragment` is dismissed, the appropriate QS focus point and snack bar get displayed based on the chosen task, and the completed tasks count is updated on the task type list item when that task is completed 
9. Repeat steps 5-8 till all tasks for a task type are complete
10. Notice that the task type is struck-through on `ImprovedMySiteFragment` once the `QuickStartFullScreenDialogFragment` is dismissed

## Regression Notes
1. Potential unintended areas of impact
    - `ImprovedMySiteFragment` -> `Quick Start - Dynamic Card` (`MySiteImprovementsFeatureConfig` - ON, `QuickStartDynamicCardsFeatureConfig` - ON )
    - `MySiteFragment` -> `Quick Start` (`MySiteImprovementsFeatureConfig` - OFF)

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Tested manually.

3. What automated tests I added (or what prevented me from doing so)
Added unit tests for actions intercepted in `ImprovedMySiteFragment` from `QuickStartFullScreenDialogFragment`. `QuickStartFullScreenDialogFragment` is shown in `MySiteImprovementsFeatureConfig` only when `QuickStartDynamicCardsFeatureConfig` - OFF.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ashiagr committed:
Restructure builder to get quick start task types using categories from the repository

ashiagr committed:
Add dismiss listener

ashiagr committed:
Tests: Update existing builder tests

ashiagr committed:
Tests: Update existing vm tests

ashiagr committed:
Tests: Add test for quick start full screen dialog dismiss action

ashiagr committed:
Add confirm listener

ashiagr committed:
Tests: Add test for quick start full screen dialog confirm action

ashiagr committed:
Tests: (Refactor) Split repository tests for completion message and dynamic card removal on refresh

ashiagr committed:
Tests: Add repository tests for refresh when dynamic card is disabled

ashiagr committed:
Don't show completion message when dynamic card disabled + same type tasks are completed

This is because
1. Both tasks types (Customise, Grow) are visible while QS is in progress and completed task type title is struck-through as a visible cue that tasks are completed
2. We call QS repository's refresh on QS full screen dialog dismiss to refresh the task types on the screen. refresh() is also called when ImprovedMySiteFragment is resumed, and since both types always exist irrespective of their completion status, isEveryQuickStartTaskDoneForType is true for the always visible completed task which triggers completion message show even on screen resume which we do not want to trigger.

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/quickstart/QuickStartBlockBuilder.kt:
```diff
@@ -1,39 +1,36 @@
-        taskTypeItems.add(buildQuickStartTaskTypeItem(localSiteId, QuickStartTaskType.GROW, onItemClick))
+        categories.forEach { category ->
+            taskTypeItems.add(buildQuickStartTaskTypeItem(category, onItemClick))
+        }
```
Suggestion (💡): What if the above was replaced with the below:
```
    fun build(
        ...
    ) = QuickStartBlock(
            taskTypeItems = categories.map { buildQuickStartTaskTypeItem(it, onItemClick) }
    )
```

ashiagr replied:
Makes it more concise 👍 
Suggestion applied in c05e241

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/quickstart/QuickStartBlockBuilder.kt:
```diff
@@ -1,39 +1,36 @@
-        val countUncompleted = quickStartStore.getUncompletedTasksByType(localSiteId, quickStartTaskType).size
+        val quickStartTaskType = category.taskType
+        val countCompleted = category.completedTasks.size
+        val countUncompleted = category.uncompletedTasks.size
```
Question (❓): So, calling `quickStartStore` is not required anymore and is removed as being redundant right? And that is done because we already have the number of completed/uncompleted tasks from the `categories` flowing from the source(s), right? I just wonder why it was implemented like this in the first place... 🤔

ashiagr replied:
> So, calling quickStartStore is not required anymore and is removed as being redundant right? And that is done because we already have the number of completed/uncompleted tasks from the categories flowing from the source(s), right? 

Right

>  I just wonder why it was implemented like this in the first place...

The logic to get completed/ uncompleted tasks count got [migrated from the `MySiteFragment` which uses `quickStartStore`](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/main/MySiteFragment.kt#L650-L665)   in [this PR](https://github.com/wordpress-mobile/WordPress-Android/pull/15141) focussing only on displaying the quick start section and not on updating it. 

With the focus on refreshing the quick start section items, this PR took this further and hooked the repository's categories (holding completed/uncompleted tasks) so that the list items can be updated using the `PartialState` logic used in `ImprovedMySiteFragment`. It also felt more natural to use the repository/ source to access the `quickStartStore` rather than the builder accessing it directy.

ParaskP7 replied:
👋 @ashiagr !

Thank you so much for the great explanation, it all makes sense to me now, great work! 🙇  🚀

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/ImprovedMySiteFragment.kt:
```diff
@@ -540,4 +547,15 @@ class ImprovedMySiteFragment : Fragment(R.layout.new_my_site_fragment),
+            val task = result.getSerializable(QuickStartFullScreenDialogFragment.RESULT_TASK) as? QuickStartTask
+            viewModel.onQuickStartFullScreenDialogConfirm(task)
+        }
+    }
```
Suggestion (💡): What if the above was replaced with the below:
```
    override fun onConfirm(result: Bundle?) {
        viewModel.onQuickStartFullScreenDialogConfirm(
                task = result?.getSerializable(QuickStartFullScreenDialogFragment.RESULT_TASK) as? QuickStartTask
        )
    }
```

ashiagr replied:
In 16fd588

ParaskP7 commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/QuickStartRepository.kt:
```diff
@@ -93,7 +95,8 @@ class QuickStartRepository
         val quickStartTaskTypes = refresh.mapAsync(coroutineScope) {
-            dynamicCardStore.getCards(siteId).dynamicCardTypes.map { it.toQuickStartTaskType() }.onEach { taskType ->
+            val quickStartTaskTypes = getQuickStartTaskTypes(siteId)
+            quickStartTaskTypes.onEach { taskType ->
```
Minor (🔍): Maybe the best equivalent to the previous piece of login is the below, wdyt?
```
getQuickStartTaskTypes(siteId).onEach { taskType ->
    ...
}
```

ashiagr replied:
In 16fd588

ParaskP7 commented on WordPress/src/test/java/org/wordpress/android/ui/mysite/QuickStartRepositoryTest.kt:
```diff
@@ -105,26 +110,59 @@ class QuickStartRepositoryTest : BaseUnitTest() {
-        quickStartRepository.refresh()
+    @Test
+    fun `given dynamic card disabled + same type tasks done, when refresh started, then completion msg not shown`() =
+            test {
```
Minor (🔍): Consider adding the below mock setup explicitly, just like it is done for when it is `true` or even for when it is `false` on some of the other such tests:

`whenever(quickStartDynamicCardsFeatureConfig.isEnabled()).thenReturn(false)`

PS: The same applies to another test further below.

ashiagr replied:
Added it explicitly in 440dc86, looks more readable, and adds to consistency 🙏.

ParaskP7 commented on WordPress/src/test/java/org/wordpress/android/ui/mysite/QuickStartRepositoryTest.kt:
```diff
@@ -105,26 +110,59 @@ class QuickStartRepositoryTest : BaseUnitTest() {
+                var result: QuickStartUpdate? = null
+                quickStartRepository.buildSource(testScope(), siteId).observeForever { result = it }
+
+                assertThat(result?.categories?.map { it.taskType }).isEqualTo(listOf(CUSTOMIZE, GROW))
```
Minor (🔍): Will the below be equivalent to this assertion here.

So, instead of setting the `result` locally:
```
var result: QuickStartUpdate? = null
quickStartRepository.buildSource(testScope(), siteId).observeForever { result = it }

assertThat(result?.categories?.map { it.taskType }).isEqualTo(listOf(CUSTOMIZE, GROW))
```

Can we do just this and reuse the field `result`:

```
assertThat(result.last().categories.map { it.taskType }).isEqualTo(listOf(CUSTOMIZE, GROW))
```

ashiagr replied:
Oh yes, it is present in the test setup and can be reused, thanks for noticing it 🙇‍♀️.
In bce72c1.

ParaskP7 reviewed (approved):
👋 @ashiagr !

I have reviewed and tested this PR as per the instructions, everything works as expected, great job! 🌟 🌟 

I have left a question (❓), some minor (🔍) comments and a couple of suggestions (💡) for you to consider. I am going to approve this PR anyway, since none is blocking. I am **NOT** going to merge this PR yet to give you some time to apply any of my suggestions. However, feel free to ignore them and merge the PR yourself.

PS: As discussed, let's create separate tasks:
1. For the additional snackbar, which is missing from the new implementation of QS, where the user can select `NOT NOW` or `GO` for when a specific QS step is finished, which is guiding them directly to the next step, without the need to go through the full-screen dialog again, just to pick it up manually.
2. For verifying that given QS is enabled through the `My Site` screen, when the app gets upgraded to the `Improved My Site` screen, then QS works as expected and can be completed even though is was started through a different implementation (`side id` vs. `selected site`).

ashiagr committed:
Make the function more concise

Transform using map + use function expression body

ashiagr committed:
Apply minor review comments

ashiagr committed:
Tests: Add mock setup explicitly instead of depending on the default one

ashiagr committed:
Tests: Reuse quick start source result from the test setup

ashiagr commented:
Thank you for a thorough review, Petros! ⭐️

I've addressed your comments and answered questions. 

Regarding 

> For the additional snackbar, which is missing from the new implementation of QS, where the user can select NOT NOW or GO for when a specific QS step is finished, which is guiding them directly to the next step, without the need to go through the full-screen dialog again, just to pick it up manually.

As discussed, "remind next task" snackbars will be added as part of this task: https://github.com/wordpress-mobile/WordPress-Android/issues/15136. I've informed Annmarie to take a look on Slack, and also added a subtask explicitly in that task. 

> For verifying that given QS is enabled through the My Site screen, when the app gets upgraded to the Improved My Site screen, then QS works as expected and can be completed even though is was started through a different implementation (side id vs. selected site).

As discussed, any issues around this part could be due to the `siteId` difference (from `AppPrefs` or `SelectedSiteRepository`). ` There's a subtask to investigate it further in the parent task #15127 for this PR.

ParaskP7 commented:
👋 @ashiagr !

Thank you so much for addressing my comments, I have now effectively resolved them all and we are once again ready for merge! 🚀 

> As discussed, "remind next task" snackbars will be added as part of this task: #15136. I've informed Annmarie to take a look on Slack, and also added a subtask explicitly in that task.

💯  🙏 

> As discussed, any issues around this part could be due to the siteId difference (from AppPrefs or SelectedSiteRepository). ` There's a subtask to investigate it further in the parent task #15127 for this PR.

💯  🙏
